[Nguyen Minh Tuan] iP

nmtuan2001 opened the pull request:
### DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's

* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Greet, Echo, Exit

Commit:
Add, List

Commit:
Mark as Done

Commit:
ToDos, Events, Deadlines

Commit:
Automated Text UI Testing

Commit:
Handle Errors

Commit:
Delete

Commit:
Delete

Commit:
Merge branch 'master' of https://github.com/nmtuan2001/ip into branch-Level-7

Commit:
Add Level-7: Save

Commit:
Add Level-8: Dates and Times

nmtuan2001 committed:
Merge branch 'branch-Level-7'

nmtuan2001 committed:
Merge branch 'branch-Level-8'

nmtuan2001 committed:
Add A-MoreOOP: Use more OOP

nmtuan2001 committed:
Add A-Packages: Organize into Packages

nmtuan2001 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

nmtuan2001 committed:
Add A-Gradle: Use Gradle

nmtuan2001 committed:
Add A-JUnit: Add JUnit Tests

nmtuan2001 committed:
Add A-JavaDoc: JavaDoc

nmtuan2001 committed:
Add A-CodingStandard: Follow the Coding Standard

nmtuan2001 committed:
Add Level-9: Find

nmtuan2001 committed:
Merge branch 'branch-A-JavaDoc'

nmtuan2001 committed:
Merge branch 'branch-A-CodingStandard'

nmtuan2001 committed:
Merge branch 'branch-Level-9'

nmtuan2001 committed:
Add A-CheckStyle: Use CheckStyle and fix style issues

leonidas1712 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,129 @@
+                            task, tasks.size());
+                    break;
+    
+                case "bye":
```
Consider adding a default branch to your switch statement as it is part of the code quality [guidelines](https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#use-the-default-branch)

leonidas1712 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,65 @@
+    public static String[] parse(String command) throws DukeException {
+        String[] splitCommands = command.split(" ", 2);
+        switch (splitCommands[0]) {
+        case "list":
+        case "sort":
+        case "bye":
```
You may want to use enums to capture the different strings that pertain to each command, in order to facilitate re-use and consistency across files that use these names (e.g Duke.java).

Alternatively, you may want to represent commands using classes for further abstraction, but this would require significantly more effort.

leonidas1712 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,94 @@
+    public void saveData(TaskList tasks) throws DukeException {
+        try {
+            FileWriter fileWriter = new FileWriter(file);
+            String content = tasks.getTasks().stream().map(x -> x.getSaveFormat() + "\n").reduce("", (x,y) -> x + y);
```
This use of declarative code with Java's functional programming features is nice!

Inserting [line breaks](https://se-education.org/guides/conventions/java/intermediate.html#layout) between chaining methods could improve readability.

leonidas1712 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,94 @@
+                default:
+                    throw new DukeException("\u2639 OOPS!!! Invalid task type: %s", type);
+                }
+                if (status.equals("1")) {
+                    task.markAsDone();
+                } else if (!status.equals("0")) {
+                    throw new DukeException("\u2639 OOPS!!! Invalid task status: %s", status);
+                }
+                tasks.add(task);
```
I like your use of flow control here to make the [happy path](https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent) obvious.

leonidas1712 commented on src/main/java/duke/task/TimedTask.java:
```diff
@@ -0,0 +1,101 @@
+     * @return Time of the task as a LocalDateTime.
+     * @throws DukeException
+     */
+    public LocalDateTime convertRawTime(String rawDateTime) throws DukeException {
+        LocalDateTime time;
+        try {
+            time = LocalDateTime.parse(rawDateTime, DateTimeFormatter.ofPattern(format));
+        } catch (DateTimeParseException e1) {
+            try {
+                time = LocalDateTime.of(LocalDate.parse(rawDateTime,
+                        DateTimeFormatter.ofPattern(format.split(" ")[0])), LocalTime.MIDNIGHT);
+            } catch (DateTimeParseException e2) {
+                try {
+                    time = LocalDateTime.of(LocalDate.now(), LocalTime.parse(rawDateTime,
+                            DateTimeFormatter.ofPattern(format.split(" ")[1])));
+                } catch (DateTimeParseException e3) {
+                    throw new DukeException("\u2639 OOPS!!! Wrong datetime format. Please input datetime in the format %s",
+                            TimedTask.format);
+                }
+            }
+        }
+        return time;
+    }
```
Perhaps this method can be re-written to avoid [deep nesting](https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting).

Correct me if I am wrong, but I think you are trying to accommodate multiple datetime formats from the user. This [SO thread ](https://stackoverflow.com/questions/4024544/how-to-parse-dates-in-multiple-formats-using-simpledateformat) could help.

nmtuan2001 replied:
Yep that's what I'm trying to do. That looks good too, I'll take a look. Thanks a lot for all your comments!

leonidas1712 reviewed (commented):
Quite well done with use of OOP, Streams and encapsulation. Using enums or more OOP for commands may help if adding more commands in the future.

nmtuan2001 committed:
Merge branch 'branch-A-CheckStyle'

nmtuan2001 committed:
Add Level-10: GUI

Sampy147 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,68 @@
+public class Ui {
+    private void print(String s) {
+        String seperator = "    ____________________________________________________________\n";
+        System.out.println(seperator + Stream.of(s.split("\n")).map(x -> "     " + x + "\n")
```
Interesting use of Streams here. Seems like a good addition as it is both intuitive and makes the code possibly less verbose

Sampy147 reviewed (commented):
Good application of OOP concepts and concepts taught in CS2030. Classes are well constructed and code is concise and easy to understand.

nmtuan2001 committed:
Fix FXML bug for GUI

nmtuan2001 committed:
Merge branch 'branch-Level-10'

nmtuan2001 committed:
Add assertions for A-Assertions

nmtuan2001 committed:
Delete redundant code and refactor for A-CodeQuality

Duke, Ui: remove code for text UI

Methods for text UI are no longer in use after GUI is set up.

Removing these methods helps clean up and significantly reduce the size of
Ui and Duke class, making it more readable.

Let's
* remove the outdated methods and keep only the method to show the
list of tasks
* remove Duke.main() and Duke.run()

Only this method is kept because it is the only one still in use.

TaskList: extract a method isSameDate

Method getTasksByDate currently contains a line using Stream.filter() that
is fairly long with multiple checks, making it hard to understand.

Extracting the section into a private helper method makes its
purpose clearer and the code more readable.

Let's add a new helper method isSameDate and break up the boolean statements
into multiple lines.

Lumping a lot of logic into one line is not desirable because it is hard
for others to understand.

nmtuan2001 committed:
Merge pull request #1 from nmtuan2001/branch-A-Assertions

Add assertions for A-Assertions

nmtuan2001 committed:
Merge branch 'master' of https://github.com/nmtuan2001/ip into branch-A-CodeQuality

nmtuan2001 committed:
Merge pull request #2 from nmtuan2001/branch-A-CodeQuality

Delete redundant code and refactor for A-CodeQuality

nmtuan2001 committed:
Set up CI

nmtuan2001 committed:
Set up CI

nmtuan2001 committed:
Fix checkstyle config

nmtuan2001 committed:
Add a way to view schedule on a specific date for B-ViewSchedules

nmtuan2001 committed:
Improve code quality by refactoring and add error handling

nmtuan2001 committed:
Add UI screenshot

nmtuan2001 committed:
Add user guide

nmtuan2001 committed:
Fix link issues in user guide
